shiny/workflowPlot/bm.R

```diff
@@ -0,0 +1,275 @@
+  if(v > 0 & m > 0){
+    output$calc_bm_button <- renderUI({actionButton("calc_bm", "Calculate Benchmarks")})
+    bm$bm_vars <- input$vars
+    bm$bm_metrics <- c()
+    if(n > 0) bm$bm_metrics <- c(bm$bm_metrics, input$metrics)
+    if(p > 0) bm$bm_metrics <- c(bm$bm_metrics, input$plots)
+  }
+  
+}, ignoreNULL = FALSE)
+
+observeEvent(input$calc_bm,{
+  req(input$all_input_id)
+  req(input$all_site_id)
+  req(bm)
+  bm$calc_bm_message <- sprintf("Setting up benchmarks")
+  output$reportvars <- renderText(paste(bm$bm_vars, seq_along(bm$bm_vars)))
+  output$reportmetrics <- renderText(paste(bm$bm_metrics))
+  
+  inputs_df <- getInputs(bety,c(input$all_site_id)) %>% 
+    dplyr::filter(input_selection_list == input$all_input_id)
+  output$inputs_df_table <- renderTable(inputs_df)
+  
+  config.list <- PEcAn.utils::read_web_config("../../web/config.php")
+  output$config_list_table <- renderTable(as.data.frame.list(config.list))
+  
+  bm$bm_settings$info <- list(userid = 1000000003) # This is my user id. I have no idea how to get people to log in to their accounts through the web interface and right now the benchmarking code has sections dependent on user id - I will fix this. 
+  bm$bm_settings$database <- list(
+    bety = list(
+      user = config.list$db_bety_username,
+      password = config.list$db_bety_password,
+      host = config.list$db_bety_hostname,
+      dbname = config.list$db_bety_database,
+      driver = config.list$db_bety_type,
+      write = TRUE
+    ),
+    dbfiles = config.list$dbfiles_folder
+  )
+  bm$bm_settings$benchmarking <- list(
+    ensemble_id = bm$ens_wf$ensemble_id,
+    new_run = FALSE
+  )
+  
+  for(i in seq_along(bm$bm_vars)){
+    benchmark <- list(
+      input_id = inputs_df$input_id,
+      variable_id = bm$bm_vars[i],
+      site_id = inputs_df$site_id,
+      metrics = list()
+    )
+    for(j in seq_along(bm$bm_metrics)){
+      benchmark$metrics = append(benchmark$metrics, list(metric_id = bm$bm_metrics[j]))
+    }
+    bm$bm_settings$benchmarking <- append(bm$bm_settings$benchmarking,list(benchmark = benchmark))
+  }
+  
+  output$calc_bm_button <- renderUI({})
+  output$print_bm_settings <- renderPrint(bm$bm_settings)
+  
+  basePath <- dplyr::tbl(bety, 'workflows') %>% dplyr::filter(id %in% bm$ens_wf$workflow_id) %>% dplyr::pull(folder)
+  
+  settings_path <- file.path(basePath, "pecan.BENCH.xml")
+  saveXML(PEcAn.utils::listToXml(bm$bm_settings,"pecan"), file = settings_path)
+  bm$settings_path <- settings_path
+  
+  bm$calc_bm_message <- sprintf("Benchmarking settings have been saved here: %s", bm$settings_path)
+  
+})
+
+# This sources the benchmarking workflow, which at the moment seems simpler than pulling functions out of it.
+# observeEvent(bm$settings_path,{ #Mostly so this is in a separate section which makes it asier to debug.
+#   settings <- PEcAn.settings::read.settings(bm$settings_path)
+#   source(system.file("scripts/benchmark.workflow.R",package = "PEcAn.benchmark"))
```

Comment:
As per previous comment, I really think it's a better design to just call the functions that register benchmarks directly, rather than calling benchmark.workflow.R.

Reply:
Agreed.
Eventually will the benchmarking workflow file itself be deprecated? Once the shiny app is done and cloning is implemented, there is no need for anyone to actually use the workflow. 
That being said, I want a version of everything that works outside of shiny (just in case we decide not to use shiny at some point...)

Reply:
I think it's fine to have the script stick around in /inst for folks who, for whatever reason, want to add benchmarks from the command line. May provide a handy example if we needed, for example, to set up a large multi-site benchmark (e.g. all of FLUXNET) to test if models capture certain emergent across-site patterns.